base64ct/src/lib.rs

```diff
@@ -138,65 +141,60 @@ const fn encoded_len_inner(n: usize, padded: bool) -> Option<usize> {
 }
 
 /// Decode the provided Base64 string into the provided destination buffer.
-#[inline]
-pub fn decode<'a>(src: &str, dst: &'a mut [u8], padded: bool) -> Result<&'a [u8], Error> {
-    let dlen = decoded_len(src, padded);
+pub fn decode(src: impl AsRef<[u8]>, dst: &mut [u8], padded: bool) -> Result<&[u8], DecodeError> {
+    let mut src = src.as_ref();
+    let mut err = 0;
+
+    if padded {
+        err = validate_padding(src)?;
+        src = &src[..unpadded_len_ct(src)];
```

Comment:
You compute `unpadded_len_ct` twice, once here and in `validate_padding`. Maybe it's worth to merge `unpadded_len_ct` and `validate_padding` into a single function?

Reply:
The only reason why I couldn't unify them into the same function is `decode_vec` needs to call `unpadded_len_ct` in isolation.

If that could be eliminated, I agree it would make more sense to combine them into the same function.

Otherwise I'm not sure how to combine them without duplication.

Reply:
You can remove padding in `decode_vec` before calling `decode`, i.e. you don't have to pass `padded` to decode.

Reply:
The over-allocation solution is an interesting one

Reply:
Okay, switched to over-allocation and unified the padding methods into `decode_padding` in a5d0bc4